R/functions-mgf.R

```diff
@@ -53,20 +67,93 @@ readMgf <- function(f, msLevel = 2L,
     ## Comment lines beginning with one of the symbols #;!/ can be
     ## included, but only outside of the BEGIN IONS and END IONS
     ## statements that delimit an MS/MS dataset.
-    cmts <- grep("^[#;!/]", mgf, perl = TRUE)
-    if (length(cmts))
-        mgf <- mgf[-cmts]
 
     begin <- grep("BEGIN IONS", mgf, fixed = TRUE) + 1L
     end <- grep("END IONS", mgf, fixed = TRUE) - 1L
-    n <- length(begin)
-    sp <- vector("list", length = n)
 
-    for (i in seq(along = sp))
-        sp[[i]] <- .extract_mgf_spectrum(mgf[begin[i]:end[i]],
-                                         mapping = mapping)
+    sp <- bplapply(seq_along(begin), FUN = function(i)
+        .extract_mgf_spectrum(mgf[begin[i]:end[i]]), BPPARAM = BPPARAM)
+
+    res <- rbindFill(sp)
```

Comment:
I tested

```r
res <- DataFrame(rbindFill(bpmapply(
    function(b, e) {
      .extract_mgf_spectrum_new(mgf = mgf[b:e], mapping = mapping)
    },
    begin,
    end,
    SIMPLIFY = FALSE,
    BPPARAM = BPPARAM
  )))
```

instead and it looks better, what do you think @jorainer?

Reply:
```r
microbenchmark::microbenchmark(
  res <- DataFrame(rbindFill(
    bplapply(
      X = seq_along(begin),
      FUN = function(i)
        .extract_mgf_spectrum(mgf[begin[i]:end[i]], mapping = mapping),
      BPPARAM = BPPARAM
    )
  )),
  
  res <- DataFrame(rbindFill(bpmapply(
    function(b, e) {
      .extract_mgf_spectrum_new(mgf = mgf[b:e], mapping = mapping)
    },
    begin,
    end,
    SIMPLIFY = FALSE,
    BPPARAM = BPPARAM
  ))),times = 5L)
```

```r
      min       lq     mean   median       uq      max neval cld
 1.935442 1.949782 2.006674 1.975322 2.080324 2.092503     5  a 
 1.545203 1.568215 1.645188 1.592491 1.645790 1.874243     5   b
```

Reply:
Well, good thing is you are right... there was a typo in the suggestion above (used `.extract_mgf_spectrum`instead of `.extract_mgf_spectrum_new` :weary: )

So the only part remaining is avoiding `sp <- ...`

Reply:
just one thing before: I will only convert the `data.frame` to a `DataFrame` at the very end, as any operation on a `DataFrame` is slower than on a `data.frame`. Also, I realized that `as(res, "DataFrame")` is faster than `DataFrame(res)`.

Reply:
Performance comparison:

```r
> microbenchmark(readMgf(f), readMgf2(f), times = 7)
Unit: seconds
        expr      min       lq     mean   median       uq      max neval cld
  readMgf(f) 17.00214 18.12558 18.45829 18.22927 19.04640 19.63269     7   a
 readMgf2(f) 16.76522 17.36399 17.91690 17.70950 18.44778 19.32001     7   a
```

So, the `mapply` is not slower. I will change that, thanks!

Reply:
hm, your benchmark puzzles me now a bit. I'll repeat the benchmark using the exact same function, just once with `bplapply` and once with `bpmapply`:

```r
> microbenchmark(readMgf_bplapply(f), readMgf_bpmapply(f), times = 7)
Unit: seconds
                expr      min       lq     mean   median       uq      max neval cld
 readMgf_bplapply(f) 29.34952 31.36238 33.65170 33.19024 35.49346 39.31048     7   a
 readMgf_bpmapply(f) 24.10515 29.81779 31.74761 32.11828 34.50458 37.36508     7   a
```

so, for this file, there is a little benefit from using `bpmapply` as you suggested.

Reply:
Repeating the benchmark led to different result:

```r
> microbenchmark(readMgf_bplapply(f), readMgf_bpmapply(f), times = 3)
Unit: seconds
                expr      min       lq     mean   median       uq      max neval cld
 readMgf_bplapply(f) 524.4349 554.9486 574.7027 585.4622 599.8366 614.2110     3   a
 readMgf_bpmapply(f) 533.6268 539.3185 556.5176 545.0102 567.9629 590.9157     3   a
```

I will run a final benchmark using 5 iterations.

Reply:
I will use `bmpapply` which avoids the additional `seq_along(begin)` call and has, as you pointed out, cleaner code.